### PR TITLE
Uncaptured aliases

### LINKED ISSUE #1879: Some aliases stopped working after commit 7be2813
I have an alias for `ranger` which is not too complicated. It does two things, it renames my tmux window and it also changes to the last navigated directory upon exiting ranger.

This alias has worked for me for a long time. After checking out commit c262e0358f053ad374e256efcd31007141eb1c7e executing `ranger` at shell was working but with the next commit, 7be281375ae74995a7878cad20f387bb1e341353, executing `ranger` hangs. If I press `Ctrl+C` I can get back to the prompt.

I would be grateful if somebody could help me identify if I'm doing something wrong or if there is a bug.

``` python
def rename_term(args, stdin=None, tcmd=None, tname=None):                                                                             
    win_info = $(tmux display-message -p) 
    if win_info:                                                                                                                      
        win_info = win_info.split()[1].rstrip(',').split(':')                                                                         
    if win_info and len(win_info) == 2:                                                                                               
        $(tmux rename-window -t @(win_info[0]) @(tname)) 
    evalx(" ".join([tcmd]+args))  #Command is actually run here
    if win_info and len(win_info) == 2:                                                                                               
        sleep 0.25                                                                                                                    
        $(tmux rename-window -t @(win_info[0]) @(win_info[1])) 

@foreground                                                                                                                           
def execRanger(args, stdin=None):                                                                                                     
    rename_term(args=args,                                                                                                            
                tcmd=$(which -s ranger).strip(),   
                tname='ranger')                                                                                                       
    with open(join(expanduser('~'), ".config/ranger/bookmarks"), 'r') as book: 
        bookmarks = book.readlines()                                                                                                  
    currDir = [line for line in bookmarks if line.startswith("'")]                                                                    
    if currDir and currDir[0]:                                                                                                        
        currDir = currDir[0][2:].replace('\n', '')                                                                                    
        cd @(currDir)                                                                                                                 

aliases['ranger'] = execRanger
```

### PR DESCRIPTION
**Pull request by scopatz**
I hope that this fixes #1879

### CODE REVIEW

File: docs/tutorial.rst [hunk 1]
```diff
@@ -1238,31 +1238,57 @@ as aliases, by wrapping them in ``@()``.  For example:
     hello world
 
 
-Foreground-only Aliases
+Unthreadable Aliases
 -----------------------
 Usually, callable alias commands will be run in a separate thread so that users
```
Comment (melund) [inline docs/tutorial.rst]: delete: users

File: docs/tutorial.rst [hunk 2]
```diff
@@ -1238,31 +1238,57 @@ as aliases, by wrapping them in ``@()``.  For example:
...
         return 'In your face!'
 
     aliases['mycmd'] = _mycmd
 
+Uncapturable Aliases
+-----------------------
+Also, callable alias by default will be executed such that their output is
+captured (like most commands in xonsh that don't enter alternate mode).
+However, some aliases may want to run alternate-mode commands themselves.
+Thus the callabale alias can't be captured without dire consequences (tm).
```
Comment (melund) [inline docs/tutorial.rst]: callabale -> callable

File: docs/tutorial.rst [hunk 3]
```diff
@@ -1238,31 +1238,57 @@ as aliases, by wrapping them in ``@()``.  For example:
...
+To prevent this, you can declare a callable alias uncapturable. This is mostly
+useful for aliases that then open up text editors, pagers, or the like.
+To make an alias run in the foreground, decorate its
```
Comment (melund) [inline docs/tutorial.rst]: replace with: To make an alias uncapturable, decorate its...

File: docs/tutorial.rst [hunk 4]
```diff
@@ -1238,31 +1238,57 @@ as aliases, by wrapping them in ``@()``.  For example:
...
-    @foreground
+    @unthreadable
     def _mycmd(args, stdin=None):
         return 'In your face!'
 
     aliases['mycmd'] = _mycmd
 
+Uncapturable Aliases
+-----------------------
+Also, callable alias by default will be executed such that their output is
```
Comment (gforsyth) [inline docs/tutorial.rst]: *alias* -> *aliases*

**Comment by gforsyth** [comment]
Hey @scopatz, thanks!  This does fix the alias issue with one caveat, if the name of the alias is the same as the program being run internally it goes wonky recursive.  That makes sense to me, but it does break from previous usage (but such is life).

```console
gil@bad_cat ~ $ from xonsh.proc import unthreadable, uncapturable
gil@bad_cat ~ $ @uncapturable
............... @unthreadable
............... def _vim(args, stdin=None):
...............     vim
...............     
gil@bad_cat ~ $ aliases['vim'] = _vim
gil@bad_cat ~ $ vim
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/home/gil/miniconda3/lib/python3.5/site-packages/xonsh/proc.py", line 1279, in wait
    r = self.f(self.args, stdin, stdout, stderr, spec)
  File "/home/gil/miniconda3/lib/python3.5/site-packages/xonsh/proc.py", line 888, in proxy_two
    return f(args, stdin)
  File "<xonsh-code>", line 4, in _vim
  File "/home/gil/miniconda3/lib/python3.5/site-packages/xonsh/built_ins.py", line 835, in subproc_captured_hiddenobject
    return run_subproc(cmds, captured='hiddenobject')
  File "/home/gil/miniconda3/lib/python3.5/site-packages/xonsh/built_ins.py", line 772, in run_subproc
    proc = spec.run(pipeline_group=pipeline_group)
  File "/home/gil/miniconda3/lib/python3.5/site-packages/xonsh/built_ins.py", line 500, in run
    self.prep_env(kwargs)
  File "/home/gil/miniconda3/lib/python3.5/site-packages/xonsh/built_ins.py", line 531, in prep_env
    denv = builtins.__xonsh_env__.detype()
  File "/home/gil/miniconda3/lib/python3.5/site-packages/xonsh/environ.py", line 751, in detype
    val = ensurer.detype(val)
  File "/home/gil/miniconda3/lib/python3.5/site-packages/xonsh/tools.py", line 854, in env_path_to_str
    return os.pathsep.join(x)
  File "/home/gil/miniconda3/lib/python3.5/_collections_abc.py", line 815, in __iter__
    v = self[i]
  File "/home/gil/miniconda3/lib/python3.5/site-packages/xonsh/tools.py", line 177, in __getitem__
    return expandpath(self._l[item])
  File "/home/gil/miniconda3/lib/python3.5/site-packages/xonsh/tools.py", line 109, in expandpath
    if env.get('EXPAND_ENV_VARS', False):
  File "/home/gil/miniconda3/lib/python3.5/site-packages/xonsh/environ.py", line 896, in get
    return self[key]
  File "/home/gil/miniconda3/lib/python3.5/site-packages/xonsh/environ.py", line 867, in __getitem__
    cabc.MutableMapping)):
  File "/home/gil/miniconda3/lib/python3.5/abc.py", line 188, in __instancecheck__
    subclass in cls._abc_negative_cache):
  File "/home/gil/miniconda3/lib/python3.5/_weakrefset.py", line 75, in __contains__
    return wr in self.data
RecursionError: maximum recursion depth exceeded in comparison
```

**Comment by scopatz** [comment]
Thanks for pointing that out @gforsyth.  I think that might be a separate issue - and it is hard to say what the correct behaviour is....(maybe it is the recursion error?)  For now though, we can avoid this issue by using `$(which -s vim)`, and because this is the only thing on the line, it explicitly needs to be in subproc mode. 

```
gil@bad_cat ~ $ from xonsh.proc import unthreadable, uncapturable
gil@bad_cat ~ $ @uncapturable
............... @unthreadable
............... def _vim(args, stdin=None):
...............     ![$(which -s vim)]
...............     
```

Thanks for the read through, both, I have made the edits requested

**Comment by gforsyth** [comment]
I'm sold.  Thanks, @scopatz!

**Pull request merged**